### DISCUSSION TITLE
How do I specify multiple return types in FastAPI with type validation?

### DISCUSSION THREAD

**Discussion by paniabhisek** [fastapi/fastapi#11409] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel

from typing import Union

app = FastAPI()

class ReturnType1(BaseModel):
    info: str = "Success message"

class ReturnType2(BaseModel):
    error: str = "Error Message"

@app.get("/{var}")
def endpoint(var: int) -> Union[ReturnType1, ReturnType2]:
    if var == 1:
        return {"info": "Success message"}
    elif var == 2:
        return {"error": "Error message"}
```

### Description

* Open the browser and call /2. It returns `{"info":"Success message"}`
* I expect it to return `{"error": "Error Message"}

### Operating System

Linux

### Operating System Details

CentOS Linux release 7.9.2009 (Core)
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

CentOS Linux release 7.9.2009 (Core)
CentOS Linux release 7.9.2009 (Core)

### FastAPI Version

0.100.0

### Pydantic Version

1.10.13

### Python Version

Python 3.9.13

**Comment by konstantinos1981** [comment]
You need to explicitly return the instance defined of your model

```python
@app.get("/{var}", response_model=Union[ReturnType1, ReturnType2])
```

**Reply by paniabhisek** [reply]
```
@app.get("/{var}", response_model = Union[ReturnType1, ReturnType2])
def endpoint(var: int):
    returntype1 = ReturnType1(info = "New success message")
    returntype2 = ReturnType2(error = "New error message")
    if var == 1:
        return returntype1
    elif var == 2:
        return returntype2
```

Above still gives "Success message" for `var == 2`.

**Reply by konstantinos1981** [reply]
seems now you redeclared...

here is the complete code
```python
from fastapi import FastAPI
from pydantic import BaseModel
from typing import Union

app = FastAPI()

class ReturnType1(BaseModel):
    info: str = "Success message"

class ReturnType2(BaseModel):
    error: str = "Error Message"

@app.get("/{var}", response_model=Union[ReturnType1, ReturnType2])
def endpoint(var: int):
    if var == 1:
        return ReturnType1()
    elif var == 2:
        return ReturnType2()
```

**Reply by paniabhisek** [reply]
I haven't redeclared. Only the instance is created. Because the return value should not depend on the pydantic model value. Pydantic model should only be used for validation and documentation. Even though, your pasted code does not work for me. It still gives "Success message" for `var == 2`.

**Reply by konstantinos1981** [reply, marked as answer]
```python
from fastapi import FastAPI
from pydantic import BaseModel
from typing import Union

app = FastAPI()

class ReturnType1(BaseModel):
    info: str = "Success message"

class ReturnType2(BaseModel):
    error: str = "Error Message"

@app.get("/{var}", response_model = Union[ReturnType1, ReturnType2])
def endpoint(var: int):
    returntype1 = ReturnType1(info="New success message")
    returntype2 = ReturnType2(error="New error message")
    if var == 1:
        return returntype1
    elif var == 2:
        return returntype2

``` 
![image](https://github.com/tiangolo/fastapi/assets/39465388/c5088bd6-213e-44f6-a0ef-14c5fb952f04)

**Reply by paniabhisek** [reply]
That is strange. I have even upgraded to the latest fasapi version and tried. No luck even by copying your code. What version you are using ? Can you please paste

**Reply by konstantinos1981** [reply]
I am running 
fastapi           0.110.1
pydantic          2.6.4
Python         3.12.2 

it should run correctly on your versions as well

**Reply by paniabhisek** [reply]
Turns out. With `pydantic == 1.10.13`, it does not work. Thanks for your help @konstantinos1981
